server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -1107,18 +1121,52 @@ public void applyClusterState(final ClusterChangedEvent event) {
         ingestClusterStateListeners.forEach(consumer -> consumer.accept(state));
 
         IngestMetadata newIngestMetadata = state.getMetadata().custom(IngestMetadata.TYPE);
-        if (newIngestMetadata == null) {
-            return;
+        if (newIngestMetadata != null) {
+            try {
+                innerUpdatePipelines(newIngestMetadata, event.state().metadata());
+            } catch (ElasticsearchParseException e) {
+                logger.warn("failed to update ingest pipelines", e);
+            }
         }
 
-        try {
-            innerUpdatePipelines(newIngestMetadata);
-        } catch (ElasticsearchParseException e) {
-            logger.warn("failed to update ingest pipelines", e);
+        updatePluginPipelines(event);
+    }
+
+    private synchronized void updatePluginPipelines(ClusterChangedEvent event) {
+
+        Map<String, List<Pipeline>> updatedPluginPipelines = new HashMap<>(pluginPipelines);
```

Comment:
How do we handle index removal?

Reply:
Index removal is handled on line 1147, when current index metadata is null 👍

Reply:
Oh I see, I missed the `indicesNames.addAll(previousIndexMetadataMap.keySet())`, thanks